src/main/scala/play/dev/filewatch/PollingFileWatchService.scala

```diff
@@ -1,103 +0,0 @@
-/*
- * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
- */
-
-package play.dev.filewatch
-
-import java.io.File
-
-import better.files.{ File => ScalaFile, _ }
-
-import annotation.tailrec
-
-/**
- * A polling Play watch service. Polls in the background.
- */
-class PollingFileWatchService(val pollDelayMillis: Int) extends FileWatchService {
-
-  def watch(filesToWatch: Seq[File], onChange: () => Unit) = {
-
-    @volatile var stopped = false
-
-    val thread = new Thread(
-      new Runnable {
-        def run() = {
-          var state = WatchState.empty
-          while (!stopped) {
-            val (triggered, newState) = SourceModificationWatch
-              .watch(() => filesToWatch.iterator.flatMap(_.toScala.listRecursively), pollDelayMillis, state)(stopped)
-            if (triggered) onChange()
-            state = newState
-          }
-        }
-      },
-      "play-watch-service"
-    )
-    thread.setDaemon(true)
-    thread.start()
-
-    new FileWatcher {
-      def stop() = stopped = true
-    }
-  }
-}
-
-/**
- * Copied from sbt.
- */
-object SourceModificationWatch {
-  type PathFinder = () => Iterator[ScalaFile]
-
-  private def listFiles(sourcesFinder: PathFinder): Set[ScalaFile] = sourcesFinder().toSet
-
-  private def findLastModifiedTime(files: Set[ScalaFile]): Long = {
-    if (files.nonEmpty) files.maxBy(_.lastModifiedTime).lastModifiedTime.toEpochMilli
-    else 0L
-  }
-
-  @tailrec def watch(sourcesFinder: PathFinder, pollDelayMillis: Int, state: WatchState)(
```

Comment:
Hmm... seems we will [not be able](https://stackoverflow.com/questions/20826786/does-java-support-and-optimize-away-tail-recursive-calls) to migrate `@tailrec` to Java here. So this is not good, I did not look into this code too much, but I think this migth be a problem when watching many files...

Reply:
No, isn't a problem. It could be if I keep implementation as recursion. But I rewrote this piece of code by simply `while` cycle that is very similar as a result bytecode after compilation `@tailrec` 😉